sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -638,73 +855,114 @@ Azure::Core::_internal::UniqueHandle<HINTERNET> WinHttpTransport::CreateRequestH
     proxyInfo.dwAccessType = WINHTTP_ACCESS_TYPE_NAMED_PROXY;
     proxyInfo.lpszProxy = const_cast<LPWSTR>(proxyWide.c_str());
     proxyInfo.lpszProxyBypass = WINHTTP_NO_PROXY_BYPASS;
-    if (!WinHttpSetOption(request.get(), WINHTTP_OPTION_PROXY, &proxyInfo, sizeof(proxyInfo)))
+    if (!WinHttpSetOption(
+            m_requestHandle.get(), WINHTTP_OPTION_PROXY, &proxyInfo, sizeof(proxyInfo)))
     {
       GetErrorAndThrow("Error while setting Proxy information.");
     }
   }
-  if (m_options.ProxyUserName.HasValue() || m_options.ProxyPassword.HasValue())
+  if (options.ProxyUserName.HasValue() || options.ProxyPassword.HasValue())
   {
     if (!WinHttpSetCredentials(
-            request.get(),
+            m_requestHandle.get(),
             WINHTTP_AUTH_TARGET_PROXY,
             WINHTTP_AUTH_SCHEME_BASIC,
-            StringToWideString(m_options.ProxyUserName.Value()).c_str(),
-            StringToWideString(m_options.ProxyPassword.Value()).c_str(),
+            StringToWideString(options.ProxyUserName.Value()).c_str(),
+            StringToWideString(options.ProxyPassword.Value()).c_str(),
             0))
     {
       GetErrorAndThrow("Error while setting Proxy credentials.");
     }
   }
 
-  if (m_options.IgnoreUnknownCertificateAuthority || !m_options.ExpectedTlsRootCertificates.empty())
+  if (options.IgnoreUnknownCertificateAuthority || !options.ExpectedTlsRootCertificates.empty())
   {
     auto option = SECURITY_FLAG_IGNORE_UNKNOWN_CA;
-    if (!WinHttpSetOption(request.get(), WINHTTP_OPTION_SECURITY_FLAGS, &option, sizeof(option)))
+    if (!WinHttpSetOption(
+            m_requestHandle.get(), WINHTTP_OPTION_SECURITY_FLAGS, &option, sizeof(option)))
     {
       GetErrorAndThrow("Error while setting ignore unknown server certificate.");
     }
   }
 
-  if (m_options.EnableCertificateRevocationListCheck)
+  if (options.EnableCertificateRevocationListCheck)
   {
     DWORD value = WINHTTP_ENABLE_SSL_REVOCATION;
-    if (!WinHttpSetOption(request.get(), WINHTTP_OPTION_ENABLE_FEATURE, &value, sizeof(value)))
+    if (!WinHttpSetOption(
+            m_requestHandle.get(), WINHTTP_OPTION_ENABLE_FEATURE, &value, sizeof(value)))
     {
       GetErrorAndThrow("Error while enabling CRL validation.");
     }
   }
 
+  // Set the callback function to be called whenever the state of the request handle changes.
+  m_httpAction = std::make_unique<_detail::WinHttpAction>(this);
+
+  if (WinHttpSetStatusCallback(
+          m_requestHandle.get(),
+          &WinHttpAction::StatusCallback,
+          WINHTTP_CALLBACK_FLAG_ALL_NOTIFICATIONS,
+          0)
+      == WINHTTP_INVALID_STATUS_CALLBACK)
+  {
+    GetErrorAndThrow("Error while setting up the status callback.");
+  }
+}
+
+/*
+ * Explicit destructor for WinHttpRequest. This is declared outside
+ * the class definition because otherwise the WinHttpAction class needs to be
+ * fully declared in win_http_transport.hpp.
```

Comment:
Why would this require `WinHttpAction` to be fully declared in win_http_transport, and why would we prefer not to do that?

Reply:
As I've mentioned before, WinHttpAction drags in a huge number of Windows-only headers which have no business being in our public API surface.

In general, I feel strongly that public headers should only declare public types. Implementation details confuse customers and it is not always clear what types are a part of the public contract and which are not.

But now that the destructor actually does something other than being an empty destructor, I'll edit the comment to make it more correct.